### PR TITLE
pid1: introduce new SERVICE_{DEAD|FAILED}_BEFORE_AUTO_RESTART service…

### PR DESCRIPTION
**Pull request by poettering**
… substates

When a service deactivates and is then automatically restarted via Restart= we currently quickly transition through
SERVICE_DEAD/SERVICE_FAILED. Which is weird given it's not the normal ("permanent") dead/failed state, but a transitory one we immediately leave from again. We do this so that software that looks for failures/successes can take notice, even if we restart as a consequence of the deactivation.

Let's clean this up a bit: let's introduce two new states: SERVICE_DEAD_BEFORE_AUTO_RESTART and SERVICE_FAILED_BEFORE_AUTO_RESTART that are used for the transitory states. Both the SERVICE_DEAD and SERVICE_DEAD_BEFORE_AUTO_RESTART will map to the high-level UNIT_INACTIVE state though. (and similar for the respective failed states). This means the high-level state machine won't change by this, only the low-level one.

This clearly seperates the substates, which makes the state engine cleaner, and allows clients to follow precisely whether we are in a transitory dead/failed state, or a permanent one, by looking at the service substate. Moreover it allows us to remove the 'n_keep_fd_store' which so far we used to ensure the fdstore was not released during this transitory dead/failed state but only during the permanent one. Since we can now distinguish these states properly we can just use that.

This has been bugging me for a while. Let's clean this up.

Note that the unit restart logic is already nicely covered in the testsiute, hence this adds no new tests for that.

And yes, this could be considered a compat break, but sofar we took the liberty to make changes to the low-level state machine (i.e. SERVICE_xyz states, sometimes called "substates") without considering this a bad breakage – the high-level state machine (i.e.  UNIT_xyz states) should be considered API that cannot be changed.

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
Is https://github.com/systemd/systemd/pull/26751 still needed with this? In other words, will dependent units still fail because of the failed transitionary state that we go through?

**Comment by poettering** [comment]
> Is #26751 still needed with this? In other words, will dependent units still fail because of the failed transitionary state that we go through?

Not sure. I am not sure I fully agree with the premise of #26751, i.e. that OnFailure= should not react on transitionary failures, only on permanent ones.

File: src/core/service.c [hunk 1]
```diff
@@ -280,24 +281,26 @@ usec_t service_restart_usec(Service *s) {
...
-
-        if (n_restarts > s->restart_steps)
-                return s->restart_usec_max;
-
-        /* Enforced in service_verify() and above */
-        assert(s->restart_usec_max > s->restart_usec);
+            s->restart_usec >= s->restart_usec_max)
+                next = s->restart_usec;
+        else if (n_restarts > s->restart_steps)
+                next =  s->restart_usec_max;
```
Comment (YHNdnzj) [inline src/core/service.c]: Duplicate space

File: src/core/service.c [hunk 2]
```diff
@@ -280,24 +281,26 @@ usec_t service_restart_usec(Service *s) {
          * between job enqueuing and running is usually neglectable compared to the time
          * we'll be sleeping. */
         n_restarts = s->n_restarts +
-                     (IN_SET(s->state, SERVICE_DEAD, SERVICE_FAILED, SERVICE_AUTO_RESTART) ? 1 : 0);
+                     (IN_SET(s->state, SERVICE_DEAD_BEFORE_AUTO_RESTART, SERVICE_FAILED_BEFORE_AUTO_RESTART, SERVICE_AUTO_RESTART) ? 1 : 0);
```
Comment (YHNdnzj) [inline src/core/service.c:284]: Hmm, if we're doing this, I think we can add 1 to it unconditionally, and change `RestartUSecCurrent` to `RestartUSecNext` instead.

If the service is running, after this change we'll still calculate the interval used last time, but logging it as "Next interval", which is spurious.

Comment (poettering) [reply src/core/service.c:284]: I wanted this PR to be as minimal as possible, the only thing other than the state engine states I added was to add debug logging here. I am think we definitely cna make further improvements on the timing calculation, but let's keep this out of this PR, to make it easily reviable.

I'll drop the word "next" from the log message for now, to make this less misleading.

Would love to review a follow-up patch adding what you are suggesting!

File: src/core/service.c [hunk 3]
```diff
@@ -3338,6 +3372,11 @@ static bool service_may_gc(Unit *u) {
             control_pid_good(s) > 0)
                 return false;
 
+        /* Only allow collection of actually dead services, i.e. not those that are in the transitionary
+         * SERVICE_DEAD_RESTART/SERVICE_FAILED_RESTART states. */
```
Comment (daandemeyer) [inline src/core/service.c]: Typos

**Review by DaanDeMeyer — APPROVED** [review]
Small nit

File: src/core/service.c (hunk 3 above)
Comment (poettering) [reply src/core/service.c]: which ones do you mean? i got the names of the service states slightly wrong. or something else?

**Pull request merged**